app/src/main/kotlin/org/fossify/keyboard/extensions/Context.kt

```diff
@@ -167,30 +167,56 @@ fun Context.setupKeyboardDialogStuff(
-        RadioItem(LANGUAGE_FRENCH_BEPO, getKeyboardLanguageText(LANGUAGE_FRENCH_BEPO)),
-        RadioItem(LANGUAGE_GERMAN, getKeyboardLanguageText(LANGUAGE_GERMAN)),
-        RadioItem(LANGUAGE_GREEK, getKeyboardLanguageText(LANGUAGE_GREEK)),
-        RadioItem(LANGUAGE_LITHUANIAN, getKeyboardLanguageText(LANGUAGE_LITHUANIAN)),
-        RadioItem(LANGUAGE_NORWEGIAN, getKeyboardLanguageText(LANGUAGE_NORWEGIAN)),
-        RadioItem(LANGUAGE_POLISH, getKeyboardLanguageText(LANGUAGE_POLISH)),
-        RadioItem(LANGUAGE_ROMANIAN, getKeyboardLanguageText(LANGUAGE_ROMANIAN)),
-        RadioItem(LANGUAGE_RUSSIAN, getKeyboardLanguageText(LANGUAGE_RUSSIAN)),
-        RadioItem(LANGUAGE_SLOVENIAN, getKeyboardLanguageText(LANGUAGE_SLOVENIAN)),
-        RadioItem(LANGUAGE_SPANISH, getKeyboardLanguageText(LANGUAGE_SPANISH)),
-        RadioItem(LANGUAGE_SWEDISH, getKeyboardLanguageText(LANGUAGE_SWEDISH)),
-        RadioItem(LANGUAGE_TURKISH_Q, getKeyboardLanguageText(LANGUAGE_TURKISH_Q)),
-        RadioItem(LANGUAGE_UKRAINIAN, getKeyboardLanguageText(LANGUAGE_UKRAINIAN)),
-        RadioItem(LANGUAGE_VIETNAMESE_TELEX, getKeyboardLanguageText(LANGUAGE_VIETNAMESE_TELEX)),
-    )
+fun Context.getKeyboardLanguages(): ArrayList<Int> {
+    val selectedLanguages = arrayListOf<Int>()
+    val selectedLanguagesFromConfig = config.selectedLanguages
+
+    for (language in selectedLanguagesFromConfig) {
+        when (language) {
+            "bengali" -> selectedLanguages.add(LANGUAGE_BENGALI)
+            "bulgarian" -> selectedLanguages.add(LANGUAGE_BULGARIAN)
+            "danish" -> selectedLanguages.add(LANGUAGE_DANISH)
+            "english_qwerty" -> selectedLanguages.add(LANGUAGE_ENGLISH_QWERTY)
+            "english_dvorak" -> selectedLanguages.add(LANGUAGE_ENGLISH_DVORAK)
+            "english_qwertz" -> selectedLanguages.add(LANGUAGE_ENGLISH_QWERTZ)
+            "french_azerty" -> selectedLanguages.add(LANGUAGE_FRENCH_AZERTY)
+            "french_bepo" -> selectedLanguages.add(LANGUAGE_FRENCH_BEPO)
+            "german" -> selectedLanguages.add(LANGUAGE_GERMAN)
+            "greek" -> selectedLanguages.add(LANGUAGE_GREEK)
+            "lithuanian" -> selectedLanguages.add(LANGUAGE_LITHUANIAN)
+            "norwegian" -> selectedLanguages.add(LANGUAGE_NORWEGIAN)
+            "polish" -> selectedLanguages.add(LANGUAGE_POLISH)
+            "romanian" -> selectedLanguages.add(LANGUAGE_ROMANIAN)
+            "russian" -> selectedLanguages.add(LANGUAGE_RUSSIAN)
+            "slovenian" -> selectedLanguages.add(LANGUAGE_SLOVENIAN)
+            "spanish" -> selectedLanguages.add(LANGUAGE_SPANISH)
+            "swedish" -> selectedLanguages.add(LANGUAGE_SWEDISH)
+            "turkish_q" -> selectedLanguages.add(LANGUAGE_TURKISH_Q)
+            "ukrainian" -> selectedLanguages.add(LANGUAGE_UKRAINIAN)
+            "vietnamese_telex" -> selectedLanguages.add(LANGUAGE_VIETNAMESE_TELEX)
+        }
```

Comment:
I think this logic can be simplified if we store the language int (e.g. `LANGUAGE_BENGALI`) in `config.selectedLanguages` instead of using language strings.

Reply:
I will check tomorrow, and see what can I do.

Reply:
@naveensingh I don't see any way to store a MutableSet<Int>, the shared preferences only have getStringSet and putStringSet.

should I convert the language's Ints to Strings and store them, then convert them to Ints on retrieval or I am missing something here?

Reply:
Let's do as you suggest and convert them to string set and back to int on retrieval. You can create helper functions inside the Config class for this.

Reply:
@naveensingh I am trying to use Gson to serialize and de-serialize the string like so:

```kotlin
    var selectedLanguages: MutableSet<Int>
        get() = Gson().fromJson(prefs.getString(SELECTED_LANGUAGES, "[]"), object: TypeToken<MutableSet<Int>>() {}.type)
        set(selectedLanguages) = prefs.edit().remove(SELECTED_LANGUAGES).putString(SELECTED_LANGUAGES, Gson().toJson(selectedLanguages)).apply()
```

I am getting this error when I click on the `Selected keyboard languages` in the settings and the app crashes:

```
java.lang.ClassCastException: java.util.HashSet cannot be cast to java.lang.String
at android.app.SharedPreferencesImpl.getString(SharedPreferencesImpl.java:310)
```

I am sorry if I am being a bit thick, but I really don't know what I am doing wrong here.

Reply:
@ronniedroid the error suggests that there is already a preference in the preferences.xml with the key "SELECTED_LANGUAGES" but it's type is HashSet which can not be cast to string now. Clearing app data should clear the existing preference and fix this `java.lang.ClassCastException`.

But we don't have to use Gson here, just do it like this:

```kotlin
var selectedLanguages: MutableSet<Int>
    get() = prefs.getStringSet(SELECTED_LANGUAGES, HashSet())!!
        .map { it.toInt() }.toMutableSet()
    set(selectedLanguages) = prefs.edit().putStringSet(
        SELECTED_LANGUAGES,
        selectedLanguages.map { it.toString() }.toSet()
    ).apply()
```

Reply:
Done, thanks